storage/innobase/handler/handler0alter.cc

```diff
@@ -3572,6 +3572,11 @@ ha_innobase::prepare_inplace_alter_table(
 		goto err_exit_no_heap;
 	}
 
+	if (srv_file_per_table && !DICT_TF_HAS_DATA_DIR(flags)
+	    && DICT_TF_HAS_DATA_DIR(prebuilt->table->flags)) {
+		flags |= (1 << DICT_TF_POS_DATA_DIR);
+	}
+
```

Comment:
I think that it would be wrong to evaluate `srv_file_per_table` multiple times. `SET GLOBAL innodb_file_per_table` may have been executed since the last time we read the variable for invoking `innobase_table_flags()`. 
If we really need that check, we should refer to `(flags2 & DICT_TF2_USE_TABLESPACE)`.

Also, do we need to test `flags` at all? Can we do just the following?
```c++
if (flags2 & DICT_TF2_USE_TABLESPACE) {
	flags |= prebuilt->table->flags & 1U << DICT_TF_POS_DATA_DIR;
}
```

Reply:
> SET GLOBAL innodb_file_per_table may have been executed since the last time we read the variable for invoking innobase_table_flags().

That means `innodb_file_per_table` should be accessed atomically.

Reply:
Yes, or the global variable `srv_file_per_table` should be read only once per ALTER TABLE statement execution. That is what my suggestion aims to do, by reading a copy of the flag from `flags2`.

Reply:
Yes, you're right: my code has such a bug.

But I meant that `srv_file_per_table` should be accessed atomically in the whole database, because it's a shared variable. Right now we have data races. This is unrelated to this MDEV.